Sources/Request/Call.swift

```diff
@@ -0,0 +1,123 @@
+	open var authenticate: ((_ request: inout URLRequest) -> Void)?
+
+	/// Initializes Call to retreive object(s) from the server.
+	/// parameter path: the path to point the call too
+	/// parameter method: the method to use for the urlRequest
+	/// parameter rootNode: used to extract JSON in method `rootNode(from:)`.
+	/// parameter serializableModel: the model is put into the body of the request as json.
+	/// parameter authenticate: optionally add authentication information to the request. Every time a request the authenticate function is called. So you can deal with authentication methods that change over time
+	public convenience init<T: Serializable> (path: String, method: HTTPMethod = .POST, rootNode: String? = nil, serializableModel: T, authenticate: ((_ request: inout URLRequest) -> Void)? = nil) {
+		self.init(path: path, method: method, rootNode: rootNode, parameter: [.jsonNode(serializableModel.json)], authenticate: authenticate)
+	}
+
+	/// Initializes Call to retreive object(s) from the server.
+	/// parameter path: the path to point the call too
+	/// parameter method: the method to use for the urlRequest
+	/// parameter rootNode: used to extract JSON in method `rootNode(from:)`.
+	/// parameter parameter: array of parameters to be added to the request when created.
+	/// parameter authenticate: optionally add authentication information to the request. Every time a request the authenticate function is called. So you can deal with authentication methods that change over time
+	public init(path: String, method: HTTPMethod = .GET, rootNode: String? = nil, parameter: [Parameter]? = nil, authenticate: ((_ request: inout URLRequest) -> Void)? = nil) {
+		self.path = path
+		self.httpMethod = method
+		self.rootNode = rootNode
+		self.parameter = parameter
+		self.authenticate = authenticate
+	}
+
+	open func request(withConfiguration configuration: Configuration) -> URLRequest? {
+		var request = URLRequest(url: URL(string: "\(configuration.baseURL)/\(path)")!, cachePolicy: .reloadIgnoringLocalAndRemoteCacheData) // uses default timeout
+		request.httpMethod = httpMethod.rawValue
+		request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+		insertParameter(request: &request)
+		authenticate?(&request)
+		return request
+	}
+
+	/// Used to begin parsing at the correct `rootnode`.
+	/// Override if you want different behaviour then:
+	/// `{"rootNode": <Any Node>}` `<Any Node> is returned when `rootNode` is set.
+	open func rootNode(from json: Any) -> JsonNode {
+		let json = extractNodeIfNeeded(from: json)
+
+		if let jsonArray = json as? [Any] {
+			return .nodeArray(jsonArray)
+		}else if let json = json as? [String: Any] {
+			return .nodeObject(json)
+		}else {
+			return .nodeNotFound(json: json ?? "")
+		}
+	}
+
+	/// Called when creating a request.
+	open func insertParameter(request: inout URLRequest) {
+		parameter?.forEach {
+			do {
+				switch $0 {
+				case .httpHeader(let headers):
```

Comment:
Je kan ook header parameters hebben die niets met authenticatie te maken hebben. Dus zit authenticatie expleciet appart ook al is het dubbel op. Maar zeker een bediscusieerbaar issue.